File: YTCnv/Helpers/UpdateChecker.cs
using CommunityToolkit.Maui.Views;
using Newtonsoft.Json;

namespace YTCnv;

public class UpdateChecker
{
    private const string GitHubApiReleases = "https://api.github.com/repos/PGAxis/YTCnv/releases/latest";

    private static SettingsSave settings = SettingsSave.Instance();

    public static bool _alreadyShown = false;

    public static async Task CheckForUpdatesAsync()
    {
        Console.WriteLine("Checking for updates...");

        try
        {
            if (settings.DontShowUpdate || _alreadyShown)
                return;

            using HttpClient client = new HttpClient();
            client.DefaultRequestHeaders.Add("User-Agent", "YTCnv-App");

            string json = await client.GetStringAsync(GitHubApiReleases);

            GitHubRelease? release = JsonConvert.DeserializeObject<GitHubRelease>(json);

            if (release == null) return;

            string latestVersion = release.tag_name.TrimStart('v');

            string[] AppVersion = AppInfo.Current.VersionString.Split('.');
            string currentVersion = $"{AppVersion[0]}.{AppVersion[1]}.{AppVersion[2]}";

            if (Version.TryParse(latestVersion, out Version latest) &&
                Version.TryParse(currentVersion, out Version current) &&
                latest > current)
            {
                await MainThread.InvokeOnMainThreadAsync(async () =>
                {
                    UpdatePopup popup = new UpdatePopup(latestVersion, release.body);
                    bool? dontShow = (bool?)await Application.Current.MainPage.ShowPopupAsync(popup);

                    if (dontShow != null)
                    {
                        if ((bool)dontShow)
                            settings.DontShowUpdate = true;
                    }
                });
            }

            _alreadyShown = true;
        }
        catch(Exception ex)
        {
            Console.WriteLine($"Update check failed: {ex.Message}");
        }
    }

    private class GitHubRelease
    {
        public string? tag_name { get; set; } = string.Empty;
        public string? body { get; set; } = string.Empty;
    }
}

File: app/src/main/java/com/pg_axis/ytcnv/UpdateChecker.kt
package com.pg_axis.ytcnv

import android.content.Context
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL

object UpdateChecker {

    private const val GITHUB_API = "https://api.github.com/repos/PGAxis/YTCnv/releases/latest"

    suspend fun checkForUpdates(context: Context, settings: ISettings): UpdateInfo? =
        withContext(Dispatchers.IO) {
            try {
                if (settings.dontShowUpdate || settings.alreadyShown) return@withContext null
                settings.alreadyShown = true

                val json = URL(GITHUB_API).openConnection().apply {
                    setRequestProperty("User-Agent", "YTCnv-App")
                }.getInputStream().bufferedReader().readText()

                val release = JSONObject(json)
                val tagName = release.getString("tag_name").trimStart('v')
                val body = release.optString("body", "")

                val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
                val currentVersion = packageInfo.versionName!!
                    .split(".")
                    .take(3)
                    .joinToString(".")

                val latest = parseVersion(tagName)
                val current = parseVersion(currentVersion)

                if (latest != null && current != null && latest > current) {
                    UpdateInfo(tagName, body)
                } else null

            } catch (e: Exception) {
                println("Update check failed: ${e.message}")
                null
            }
        }

    private fun parseVersion(version: String): Triple<Int, Int, Int>? {
        val parts = version.split(".").mapNotNull { it.toIntOrNull() }
        return if (parts.size >= 3) Triple(parts[0], parts[1], parts[2]) else null
    }

    private fun isInstalledFromStore(context: Context): Boolean {
        val installer = context.packageManager
            .getInstallSourceInfo(context.packageName)
            .installingPackageName

        return installer == "org.fdroid.fdroid" ||
                installer == "com.aurora.store" ||
                installer == "org.gdroid.gdroid"
    }

    private operator fun Triple<Int, Int, Int>.compareTo(other: Triple<Int, Int, Int>): Int {
        if (first != other.first) return first - other.first
        if (second != other.second) return second - other.second
        return third - other.third
    }
}

data class UpdateInfo(val version: String, val changelog: String)